### PR TITLE
[refactor] make methods private in macros.jl

### PR DESCRIPTION
**Pull request by odow**
Part of #1790

What should we do with extendable methods like `_build_constraint` and `_build_variable`?

### CODE REVIEW

**Comment by blegat** [comment]
How does one distinguish between a method meant to be extended and an internal method ? Maybe we shouldn't underscore functions of the JuMP extension API.

**Comment by blegat** [comment]
Note that it is perfectly fine to create constraint with build_constraint and add_constraint. The @constraint macro simply adds convenience. So it is not like private functions that the user shouldn't use

**Comment by blegat** [comment]
> Will you commit to not break anything between minor JuMP releases? If so and if the function is documented then it can be exported.

Certainly, that would allow JuMP extension to safely use e.g. `JuMP 0.19 0.20` in their `REQUIRE` to prevent breakage.

> I suggest we keep them as internal features for now, and export them in a future release once we're convinced that the API is stable.

I am not excited to go back and forth with renaming with corresponding updates to extensions, clashes if the wrong branches are checkout... I would prefer keeping extension functions exported.

**Comment by mlubin** [comment]
@blegat effectively owns this part of the API, so I'll leave the decision on underscoring to him.

**Comment by odow** [comment]
@blegat, can you make a list of extension functions you consider to be part of the public API? I restored `build_constraint` and `parse_constraint`.

File: src/macros.jl [hunk 1]
```diff
@@ -339,13 +339,13 @@ end
...
 # two-argument build_constraint is used for one-sided constraints.
 # Right-hand side is zero.
-sense_to_set(_error::Function, ::Union{Val{:(<=)}, Val{:(≤)}}) = MOI.LessThan(0.0)
-sense_to_set(_error::Function, ::Union{Val{:(>=)}, Val{:(≥)}}) = MOI.GreaterThan(0.0)
-sense_to_set(_error::Function, ::Val{:(==)}) = MOI.EqualTo(0.0)
-sense_to_set(_error::Function, ::Val{S}) where S = _error("Unrecognized sense $S")
+_sense_to_set(_error::Function, ::Union{Val{:(<=)}, Val{:(≤)}}) = MOI.LessThan(0.0)
+_sense_to_set(_error::Function, ::Union{Val{:(>=)}, Val{:(≥)}}) = MOI.GreaterThan(0.0)
+_sense_to_set(_error::Function, ::Val{:(==)}) = MOI.EqualTo(0.0)
+_sense_to_set(_error::Function, ::Val{S}) where S = _error("Unrecognized sense $S")
```
Comment (blegat) [inline src/macros.jl]: I would leave this one exported

File: src/macros.jl [hunk 2]
```diff
@@ -339,13 +339,13 @@ end
...
 
-function parse_one_operator_constraint(_error::Function, vectorized::Bool,
-                                       ::Union{Val{:in}, Val{:∈}}, aff, set)
+function _parse_one_operator_constraint(_error::Function, vectorized::Bool,
```
Comment (blegat) [inline src/macros.jl]: This one too

File: src/macros.jl [hunk 3]
```diff
@@ -356,19 +356,19 @@ function parse_one_operator_constraint(_error::Function, vectorized::Bool,
...
 end
 
 function parse_constraint(_error::Function, sense::Symbol, lhs, rhs)
     (sense, vectorized) = _check_vectorized(sense)
-    vectorized, parse_one_operator_constraint(_error, vectorized, Val(sense), lhs, rhs)...
+    vectorized, _parse_one_operator_constraint(_error, vectorized, Val(sense), lhs, rhs)...
 end
 
-function parse_ternary_constraint(_error::Function, vectorized::Bool, lb, ::Union{Val{:(<=)}, Val{:(≤)}}, aff, rsign::Union{Val{:(<=)}, Val{:(≤)}}, ub)
+function _parse_ternary_constraint(_error::Function, vectorized::Bool, lb, ::Union{Val{:(<=)}, Val{:(≤)}}, aff, rsign::Union{Val{:(<=)}, Val{:(≤)}}, ub)
```
Comment (blegat) [inline src/macros.jl]: This one too

File: src/macros.jl [hunk 4]
```diff
@@ -563,28 +563,28 @@ function constraint_macro(args, macro_name::Symbol, parsefun::Function)
...
+        macro_code = _macro_assign_and_return(creationcode, variable, name,
                                              model_for_registering = m)
     end
-    return assert_validmodel(m, macro_code)
+    return _assert_valid_model(m, macro_code)
 end
 
 # This function needs to be implemented by all `AbstractModel`s
-constraint_type(m::Model) = ConstraintRef{typeof(m)}
+_constraint_type(m::Model) = ConstraintRef{typeof(m)}
```
Comment (blegat) [inline src/macros.jl]: This one too

File: src/macros.jl [hunk 5]
```diff
@@ -995,44 +982,42 @@ macro expression(args...)
...
+_esc_non_constant(x::Expr) = isexpr(x,:quote) ? x : esc(x)
+_esc_non_constant(x) = esc(x)
 
-# Returns the type of what `add_variable(::Model, build_variable(...))` would return where `...` represents the positional arguments.
-# Example: `@variable m [1:3] foo` will allocate an vector of element type `variable_type(m, foo)`
+# Returns the type of what `add_variable(::Model, _build_variable(...))` would return where `...` represents the positional arguments.
+# Example: `@variable m [1:3] foo` will allocate an vector of element type `_variable_type(m, foo)`
 # Note: it needs to be implemented by all `AbstractModel`s
-variable_type(m::Model) = VariableRef
+_variable_type(m::Model) = VariableRef
```
Comment (blegat) [inline src/macros.jl]: This one too

File: src/macros.jl [hunk 6]
```diff
@@ -995,44 +982,42 @@ macro expression(args...)
...
 # Returns a new variable. Additional positional arguments can be used to dispatch the call to a different method.
 # The return type should only depends on the positional arguments for `variable_type` to make sense. See the @variable macro doc for more details.
-# Example: `@variable m x` foo will call `build_variable(_error, info, foo)`
-function build_variable(_error::Function, info::VariableInfo; extra_kw_args...)
+# Example: `@variable m x` foo will call `_build_variable(_error, info, foo)`
+function _build_variable(_error::Function, info::VariableInfo; extra_kw_args...)
```
Comment (blegat) [inline src/macros.jl]: This one too

File: src/macros.jl [hunk 7]
```diff
@@ -1048,21 +1033,21 @@ function namecall(base_name, idxvars)
...
-reverse_sense(::Val{:(==)}) = Val(:(==))
+_reverse_sense(::Val{:<=})   = Val(:>=)
+_reverse_sense(::Val{:≤})    = Val(:≥)
+_reverse_sense(::Val{:>=})   = Val(:<=)
+_reverse_sense(::Val{:≥})    = Val(:≤)
+_reverse_sense(::Val{:(==)}) = Val(:(==))
 
 """
-    parse_one_operator_variable(_error::Function, infoexpr::VariableInfoExpr, sense::Val{S}, value) where S
+    _parse_one_operator_variable(_error::Function, infoexpr::VariableInfoExpr, sense::Val{S}, value) where S
```
Comment (blegat) [inline src/macros.jl]: All the `parse_variable` API too

**Comment by odow** [comment]
Can we have a rule for which ones should and shouldn't be? A lot of these are undocumented... Maybe you want to take over this PR and export the ones you want and add docstrings to them? I'm flying blind here...

**Comment by blegat** [comment]
All the parse/build/add API should the exported and also variable_type and constraint_type since it is used to extend the variable and constraint macros.
I can take over on Sundays but not before

**Comment by odow** [comment]
Can we make a rule that all exported functions need to have docstrings? For example, why is `sense_to_set` exported? What is the use-case for overloading this?

**Comment by blegat** [comment]
> Can we make a rule that all exported functions need to have docstrings?

I have created an issue. Edit the issue if I missed anything.

> For example, why is sense_to_set exported? What is the use-case for overloading this?

Defining the meaning of a new sense symbol.

**Pull request merged**